sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala

```diff
@@ -0,0 +1,139 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet
+
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, Expression, NamedExpression}
+import org.apache.spark.sql.catalyst.planning.{PhysicalOperation, ProjectionOverSchema, SelectedField}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Prunes unnecessary Parquet columns given a [[PhysicalOperation]] over a
+ * [[ParquetRelation]]. By "Parquet column", we mean a column as defined in the
+ * Parquet format. In Spark SQL, a root-level Parquet column corresponds to a
+ * SQL column, and a nested Parquet column corresponds to a [[StructField]].
+ */
+private[sql] object ParquetSchemaPruning extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan =
+    plan transformDown {
+      case op @ PhysicalOperation(projects, filters,
+          l @ LogicalRelation(hadoopFsRelation @ HadoopFsRelation(_, partitionSchema,
+            dataSchema, _, parquetFormat: ParquetFileFormat, _), _, _)) =>
+        val projectionFields = projects.flatMap(getFields)
+        val filterFields = filters.flatMap(getFields)
+        val requestedFields = (projectionFields ++ filterFields).distinct
+
+        // If [[requestedFields]] includes a proper field, continue. Otherwise,
+        // return [[op]]
+        if (requestedFields.exists { case (_, optAtt) => optAtt.isEmpty }) {
+          val prunedSchema = requestedFields
+            .map { case (field, _) => field }
+            .map(field => StructType(Array(field)))
+            .reduceLeft(_ merge _)
+          val parquetDataColumnNames = dataSchema.fieldNames
+          val prunedDataSchema =
+            StructType(prunedSchema.filter(f => parquetDataColumnNames.contains(f.name)))
+          val parquetDataFields = dataSchema.fields.toSet
+          val prunedDataFields = prunedDataSchema.fields.toSet
+
+          // If the original Parquet relation data fields are different from the
+          // pruned data fields, continue. Otherwise, return [[op]]
+          if (parquetDataFields != prunedDataFields) {
+            val dataSchemaFieldNames = hadoopFsRelation.dataSchema.fieldNames
+            val newDataSchema =
+              StructType(prunedSchema.filter(f => dataSchemaFieldNames.contains(f.name)))
```

Comment:
Is `newDataSchema` actually `prunedDataSchema`, if `dataSchemaFieldNames` is `parquetDataColumnNames`?